sticker/src/tensorflow/dataset.rs

```diff
@@ -209,3 +231,51 @@ where
         None
     }
 }
+
+/// An Iterator adapter performing local shuffling.
+///
+/// Fills a buffer with size `buffer_size` on the first call. Subsequent
+/// calls swap the next incoming item with a random element from the
+/// buffer and return the random element.
+pub struct Shuffled<I> {
+    inner: I,
+    buffer: Vec<Result<Sentence, Error>>,
+    buffer_size: usize,
+    rng: XorShiftRng,
+}
+
+impl<I> Iterator for Shuffled<I>
+where
+    I: Iterator<Item = Result<Sentence, Error>>,
+{
+    type Item = Result<Sentence, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.buffer.is_empty() {
```

Comment:
`if self.buffer.len() < self.buffer_size` is probably more correct here (especially when failing eagerly as suggested in the other comment).

Reply:
That'd mean we go `self.buffer_size  - 1` times into this if-branch to check whether `self.inner` has something to offer.

Reply:
Why would you go that number of times into the `if` branch? Only the first time, right, when the buffer is filled?

Reply:
If `self.inner` is exhausted, `self.buffer.len()`will be smaller than `self.buffer_size` for the remaining calls to `self.next`.

Reply:
Ah, sorry, I misread it as going into the branch for the the overall number of items - 1, misread this! So looks fine then!

Reply:
So keep `self.buffer.is_empty`?